src/test/java/fi/metatavu/metaform/server/test/functional/tests/TemplateTestsIT.kt

```diff
@@ -0,0 +1,412 @@
+                template = updatedTemplate
+        )
+
+        assertEquals(2, testBuilder.systemAdmin.templates.list(TemplateVisibility.PUBLIC).size)
+
+    }
+
+    @Test
+    @Throws(Exception::class)
+    fun testDeleteTemplate() = TestBuilder().use { testBuilder ->
+
+        val createdMetaform = testBuilder.systemAdmin.metaforms.createFromJsonFile("simple")
+        assertNotNull(createdMetaform)
+
+        val template1 = createTemplateFromMetaform(
+                metaform = createdMetaform,
+                templateVisibility = TemplateVisibility.PUBLIC
+        )
+
+        val createdTemplate = testBuilder.systemAdmin.templates.createTemplate(template = template1)
+
+        assertNotNull(createdTemplate)
+        assertNotNull(createdTemplate.id)
+
+        val foundTemplate = testBuilder.systemAdmin.templates.findTemplate(templateId = createdTemplate.id!!)
+
+        assertNotNull(foundTemplate)
+        assertNotNull(foundTemplate.id)
+        testBuilder.systemAdmin.templates.delete(templateId = foundTemplate.id!!)
+        testBuilder.systemAdmin.templates.assertFindFailStatus(404, templateId = foundTemplate.id)
+
+    }
+
+    @Test
+    @Throws(Exception::class)
+    fun testUpdateTemplate() = TestBuilder().use { testBuilder ->
+
+        val simpleMetaform: Metaform = testBuilder.systemAdmin.metaforms.createFromJsonFile("simple")
+        assertNotNull(simpleMetaform)
+        assertNotNull(simpleMetaform.id)
+        val tbncMetaform: Metaform = testBuilder.systemAdmin.metaforms.createFromJsonFile("tbnc")
+        assertNotNull(tbncMetaform)
+        assertNotNull(tbncMetaform.id)
+
+        val createdTemplate = testBuilder.systemAdmin.templates.createTemplate(
+                template = createTemplateFromMetaform(
+                        metaform = simpleMetaform,
+                        templateVisibility = TemplateVisibility.PRIVATE
+                )
+        )
+
+        assertNotNull(createdTemplate)
+        assertNotNull(createdTemplate.id)
+        assertEquals(TemplateVisibility.PRIVATE, createdTemplate.visibility)
+
+        //lets change some existing data and update it to database
+        val changedTemplate: Template = createdTemplate.copy(
+                visibility = TemplateVisibility.PUBLIC,
+                data = TemplateData(
+                        title = tbncMetaform.title,
+                        exportThemeId = tbncMetaform.exportThemeId,
+                        sections = tbncMetaform.sections
+                )
+        )
+
+        val updatedTemplate: Template = testBuilder.systemAdmin.templates.updateTemplate(
+            id = createdTemplate.id!!,
+            template = changedTemplate
+        )
+
+        assertNotNull(updatedTemplate)
+        assertNotNull(updatedTemplate.id)
+        assertNotNull(updatedTemplate.visibility)
+
+        assertEquals(updatedTemplate.id, changedTemplate.id)
+        assertEquals(updatedTemplate.visibility, changedTemplate.visibility)
```

Comment:
Missing assert that changed template actually is changed. e.g.
 
      testBuilder.systemAdmin.templates.assertSectionEqual(
            expected = testTemplate.data.sections?.get(0),
            actual = changedTemplate.data.sections?.get(0)
      )